test/Path/Gen.hs

```diff
@@ -1,56 +1,109 @@
+--
+-- * Its path is a relative path
+-- * Its path does not have a trailing path separator
+-- * Its path is valid according to 'System.FilePath's definition.
+-- * Its path is not '.'
+-- * Its path is not empty
+-- * Its path does not end in '/.'
+-- * Its path is not '.'
+-- * Its path does not contain '..'.
+-- * Parsing the path and rendering it again results in the same path.
 instance Validity (Path Rel File) where
-  isValid p@(Path fp)
-    =  FilePath.isRelative fp
-    && not (FilePath.hasTrailingPathSeparator fp)
-    && FilePath.isValid fp
-    && fp /= "."
-    && not (hasParentDir fp)
-    && (parseRelFile fp == Just p)
-
+  validate p@(Path fp) =
+    mconcat
+      [ declare "The path is relative." $ FilePath.isRelative fp
+      , declare "The path has no trailing path separator." $
+        not (FilePath.hasTrailingPathSeparator fp)
+      , declare "System.FilePath considers the path valid." $ FilePath.isValid fp
+      , declare "The path does not equal \".\"" $ fp /= "."
+      , declare "The path is not empty" $ not (null fp)
+      , declare "The path does not end in /." $ not ("/." `isSuffixOf` fp)
+      , declare "The path does not a parent directory." $ not (hasParentDir fp)
+      , declare "The path can be identically parsed as a relative file path." $
+        parseRelFile fp == Just p
+      ]
+
+-- | An absolute path to a directory is valid if:
+--
+-- * Its path is an absolute path
+-- * Its path has a trailing path separator
+-- * Its path is valid according to 'System.FilePath's definition.
+-- * Its path does not contain '..'.
+-- * Parsing the path and rendering it again results in the same path.
 instance Validity (Path Abs Dir) where
-  isValid p@(Path fp)
-    =  FilePath.isAbsolute fp
-    && FilePath.hasTrailingPathSeparator fp
-    && FilePath.isValid fp
-    && not (hasParentDir fp)
-    && (parseAbsDir fp == Just p)
-
+  validate p@(Path fp) =
+    mconcat
+      [ declare "The path is absolute." $ FilePath.isAbsolute fp
+      , declare "The path has a trailing path separator." $ FilePath.hasTrailingPathSeparator fp
+      , declare "System.FilePath considers the path valid." $ FilePath.isValid fp
+      , declare "The path does not a parent directory." $ not (hasParentDir fp)
+      , declare "The path can be identically parsed as an absolute directory path." $
+        parseAbsDir fp == Just p
+      ]
+
+-- | A relative path to a directory is valid if:
+--
+-- * Its path is a relative path
+-- * Its path has a trailing path separator
+-- * Its path is valid according to 'System.FilePath's definition.
+-- * Its path does not contain '..'.
+-- * Parsing the path and rendering it again results in the same path.
 instance Validity (Path Rel Dir) where
-  isValid p =
-    FilePath.isRelative fp
-    && FilePath.hasTrailingPathSeparator fp
-    && FilePath.isValid fp
-    && not (hasParentDir fp)
-    && (parseRelDir fp == Just p)
-    where fp = toFilePath p
+  validate (Path "./") = valid
```

Comment:
This is not right. We cannot have "./" in `Path`, "." is represented as an empty string "".